src/plugins/data/public/ui/search_bar_extensions/search_bar_extension.tsx

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { EuiErrorBoundary } from '@elastic/eui';
+import React, { useEffect, useMemo, useRef, useState } from 'react';
+import ReactDOM from 'react-dom';
+import { IIndexPattern } from '../../../common';
+import { DataSource } from '../../data_sources/datasource';
+
+interface SearchBarExtensionProps {
+  config: SearchBarExtensionConfig;
+  dependencies: SearchBarExtensionDependencies;
+  componentContainer: Element;
+  bannerContainer: Element;
+}
+
+export interface SearchBarExtensionDependencies {
+  /**
+   * Currently selected index patterns.
+   */
+  indexPatterns?: Array<IIndexPattern | string>;
+  /**
+   * Currently selected data source.
+   */
+  dataSource?: DataSource;
+  /**
+   * Currently selected query language.
+   */
+  language: string;
+}
+
+export interface SearchBarExtensionConfig {
+  /**
+   * The id for the search bar extension.
+   */
+  id: string;
+  /**
+   * Lower order indicates higher position on UI.
+   */
+  order: number;
+  /**
+   * A function that determines if the search bar extension is enabled and should be rendered on UI.
+   * @returns whether the extension is enabled.
+   */
+  isEnabled: (dependencies: SearchBarExtensionDependencies) => Promise<boolean>;
+  /**
+   * A function that returns the search bar extension component. The component
+   * will be displayed on top of the query editor in the search bar.
+   * @param dependencies - The dependencies required for the extension.
+   * @returns The component the search bar extension.
+   */
+  getComponent?: (dependencies: SearchBarExtensionDependencies) => React.ReactElement | null;
```

Comment:
just noting here so don't change anything but i wonder if we should be consist with the return when related `undefined` or null.

https://github.com/opensearch-project/OpenSearch-Dashboards/pull/7001#discussion_r1637006688

any thoughts or preference? like i said, don't modify this and i didn't really have a good reason for undefined either so any opinion will be cool. if don't have a chance then that's cool.

Reply:
react functional component type definition uses null so i put null
```ts
    interface FunctionComponent<P = {}> {
        (props: PropsWithChildren<P>, context?: any): ReactElement<any, any> | null;
        propTypes?: WeakValidationMap<P> | undefined;
        contextTypes?: ValidationMap<any> | undefined;
        defaultProps?: Partial<P> | undefined;
        displayName?: string | undefined;
    }
```